### ISSUE TITLE
`ValueError: oauth_client_id cannot be empty` exception with custom notebook image

### ISSUE THREAD

**Issue by mrocklin** [jupyterhub/jupyterhub#1571]
Howdy folks, 

I'm trying to stand up JupyterHub (Helm chart - edited by @willingc) version v0.4 with a custom notebook image.  Everything works fine with the base-notebook image but fails with my custom image.  I don't know why and was hoping that you all could help out here.

**How to reproduce the issue**

I'm going through the Z2JH instructions

```yaml
proxy:
  secretToken: XXXX
hub:
  cookieSecret: XXXX

singleuser:
  image: 
      name: mrocklin/dask-notebook
      tag: 2
      extraEnv:
        EXTRA_PIP_PACKAGES: gcsfs
```

    helm install --name=jhub jupyterhub/jupyterhub --version=v0.4 -f config.yaml

With this docker image

### Dockerfile

```docker
FROM jupyter/base-notebook

RUN conda install --yes -c conda-forge \
    dask distributed numpy pandas numba nomkl fastparquet zict blosc \
    cytoolz ipywidgets \
    && conda clean -tipsy
RUN jupyter labextension install @jupyter-widgets/jupyterlab-manager
```

**What actually happens**

On the web client I get a 500 error.  Looking at the logs in Kubernetes I get the following traceback:

<details>

```python
+ '[' -e /opt/app/environment.yml ']'
+ echo 'no environment.yml'
+ '[' '' ']'
+ '[' '' ']'
+ jupyterhub-singleuser '--user="rocklin"' '--cookie-name="jupyter-hub-token-rocklin"' '--base-url="/user/rocklin"' '--hub-host=""' '--hub-prefix="/hub/"' '--hub-api-url="http://10.11.244.166:8081/hub/api"' '--ip="0.0.0.0"' --port=8888
no environment.yml
[W 2017-12-04 22:22:16.189 rocklin configurable:168] Config option `open_browser` not recognized by `SingleUserNotebookApp`.  Did you mean `browser`?
/opt/conda/lib/python3.6/site-packages/traitlets/config/configurable.py:84: UserWarning: Config option `cookie_name` not recognized by `HubOAuth`.
  self.config = config
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 528, in get
    value = obj._trait_values[self.name]
KeyError: 'oauth_client_id'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/bin/jupyterhub-singleuser", line 6, in <module>
    main()
  File "/opt/conda/lib/python3.6/site-packages/jupyterhub/singleuser.py", line 455, in main
    return SingleUserNotebookApp.launch_instance(argv)
  File "/opt/conda/lib/python3.6/site-packages/jupyter_core/application.py", line 266, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/traitlets/config/application.py", line 657, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-7>", line 2, in initialize
  File "/opt/conda/lib/python3.6/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/notebook/notebookapp.py", line 1368, in initialize
    self.init_webapp()
  File "/opt/conda/lib/python3.6/site-packages/jupyterhub/singleuser.py", line 393, in init_webapp
    self.init_hub_auth()
  File "/opt/conda/lib/python3.6/site-packages/jupyterhub/singleuser.py", line 388, in init_hub_auth
    if not self.hub_auth.oauth_client_id:
  File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 556, in __get__
    return self.get(obj, cls)
  File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 535, in get
    value = self._validate(obj, dynamic_default())
  File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 593, in _validate
    value = self._cross_validate(obj, value)
  File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 599, in _cross_validate
    value = obj._trait_validators[self.name](obj, proposal)
  File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 907, in __call__
    return self.func(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/jupyterhub/services/auth.py", line 439, in _ensure_not_empty
    raise ValueError("%s cannot be empty." % proposal.trait.name)
ValueError: oauth_client_id cannot be empty.
```

This looks similar to https://github.com/jupyterhub/wrapspawner/issues/11

**Comment by ziedbouf** [comment]
Can you provide more details on jupyterhub extension inside the docker image?

**Comment by mrocklin** [comment]
It is the ipywidgets extension described here: http://ipywidgets.readthedocs.io/en/stable/user_install.html#installing-the-jupyterlab-extension

**Comment by ziedbouf** [comment]
No can you access the started docker images and check the `jupyterhub --version` because i got this error due to difference between jupyterhub extension on the host and the one on the docker container

**Comment by mrocklin** [comment]
I don't know what you mean by jupyterhub extension

**Comment by mrocklin** [comment]
Hrm, I'm not sure I can even inspect the docker image.  It doesn't seem to start up.  Also, I'm surprised that  `jupyterhub` would be running on the user's image.

**Comment by choldgraf** [comment]
I believe the base-notebook image runs jupyterhub 0.8: https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile#L79

Could you try downgrading jupyterhub to 0.7 and see if that changes anything @mrocklin ? I _think_ that the jupyterhub version needs to be the same in the image and in the kubernetes deployment. v0.4 of the helm chart uses jupyterhub 0.7 IIRC

**Comment by ziedbouf** [comment]
That's my intution at the first time i seen't. you are expecting jupyterlab but it seems that jupyterhub already exist on the docker images to spin up (in singleuser mode) from` the hub if you are using it.

**Comment by willingc** [comment]
@mrocklin Perhaps creating the Dockerfile using repo2docker such as described [here](https://zero-to-jupyterhub.readthedocs.io/en/v0.4-doc/user-experience.html#build-a-custom-image-with-repo2docker) using an `environment.yml` for the conda and extra pip dependency and using `postBuild` to run the jupyterlab step. @minrk thoughts?

**Comment by ziedbouf** [comment]
This is an images i created when using jupyterhub from `https://github.com/natbusa/jupyterhub-ansible-deploy` by just removing jupyterlab and reinstall the needed version of jupyterhub.
```# Build as natbusa/jupyterhub-scipy-lab
# Run with the DockerSpawner in JupyterHub

FROM jupyter/scipy-notebook

MAINTAINER xxx  <xxxx@hey>

USER root

# delete the default user, create the user while starting jupyterhub-singleuser via the CMD script
RUN userdel jovyan && rm -rf /home/jovyan
ADD systemuser.sh /srv/singleuser/systemuser.sh

# Override jupyter notebook config file
RUN rm -f  /etc/jupyter/jupyter_notebook_config.py


# Remove jupyterlab from the imamges
RUN conda uninstall jupyterlab jupyterlab_launcher -y

# Remove jupyterhub 0.8.1
RUN conda uninstall jupyterhub  -y

# Install jupyterub 0.7.2
RUN conda install jupyterhub==0.7.2 -y

# smoke test entrypoint
RUN USER_ID=65000 USER=systemusertest sh /srv/singleuser/systemuser.sh -h && userdel systemusertest

CMD ["sh", "/srv/singleuser/systemuser.sh"]
```

**Comment by minrk** [comment]
This is a jupyterhub version incompatibility. Either trying out the v0.5 helm chart (just released! Has jupyterhub 0.8) or install jupyterhub 0.7 in your user image to work with the 0.4 chart.

This will generally do it:

```Dockerfile
RUN conda install -y jupyterhub==0.7.2 # or pip
```

**Comment by mrocklin** [comment]
Verified that this was the issue.   Thanks all for the prompt help and useful suggestions!  Closing
